src/qibojit/custom_operators/__init__.py

```diff
@@ -10,20 +10,23 @@ class CupyCpuDevice:  # pragma: no cover
 
     def __init__(self, K):
         self.K = K
+        self.original_engine = K.engine.name
 
     def __enter__(self, *args):
         self.K.set_engine("numba")
 
     def __exit__(self, *args):
         if self.K.gpu_devices:
-            self.K.set_engine("cupy")
+            self.K.set_engine(self.original_engine)
 
 
 class JITCustomBackend(NumpyBackend, AbstractCustomOperators):
 
     description = "Uses custom operators based on numba.jit for CPU and " \
                   "custom CUDA kernels loaded with cupy GPU."
 
+    default_gpu_engine = "cuquantum"
```

Comment:
I believe the current cupy backend should remain as default for GPU until we have enough evidence (through benchmarks and tests) that cuquantum works correctly and with at least equivalent performance. I wouldn't be surprised if cuquantum is slightly slower for some gates such as SWAP since it is using matrix multiplication.

That being said, we could consider adding an additional mechanism to switch between these backends. Currently the switch is possible via qibo using:
```Python
from qibo import K
K.set_engine("cuquantum")
```
which is I believe is acceptable as API, as long as we document it somewhere. Perhaps we could additionally modify the default GPU engine via an environment variable, instead of hard-coding it in the code.

Reply:
I reset cupy as the default backend engine for gpu.
I also think that the `set_engine` method is an acceptable API for now.